// rsvp/lib/rsvp.js
import Promise from 'rsvp/lib/rsvp/promise';
import EventTarget from 'rsvp/lib/rsvp/events';
import denodeify from 'rsvp/lib/rsvp/node';
import all from 'rsvp/lib/rsvp/all';
import allSettled from 'rsvp/lib/rsvp/all-settled';
import race from 'rsvp/lib/rsvp/race';
import hash from 'rsvp/lib/rsvp/hash';
import hashSettled from 'rsvp/lib/rsvp/hash-settled';
import rethrow from 'rsvp/lib/rsvp/rethrow';
import defer from 'rsvp/lib/rsvp/defer';
import {
  config,
  configure
} from 'rsvp/lib/rsvp/config';
import map from 'rsvp/lib/rsvp/map';
import resolve from 'rsvp/lib/rsvp/resolve';
import reject from 'rsvp/lib/rsvp/reject';
import filter from 'rsvp/lib/rsvp/filter';
import asap from 'rsvp/lib/rsvp/asap';

// defaults
config.async = asap;
config.after = cb => setTimeout(cb, 0);
const cast = resolve;

const async = (callback, arg) => config.async(callback, arg);

function on() {
  config['on'].apply(config, arguments);
}

function off() {
  config['off'].apply(config, arguments);
}

// Set up instrumentation through `window.__PROMISE_INTRUMENTATION__`
if (typeof window !== 'undefined' && typeof window['__PROMISE_INSTRUMENTATION__'] === 'object') {
  let callbacks = window['__PROMISE_INSTRUMENTATION__'];
  configure('instrument', true);
  for (let eventName in callbacks) {
    if (callbacks.hasOwnProperty(eventName)) {
      on(eventName, callbacks[eventName]);
    }
  }
}

import platform from 'rsvp/lib/rsvp/platform';
// the default export here is for backwards compat:
//   https://github.com/tildeio/rsvp.js/issues/434
export default {
  asap,
  cast,
  Promise,
  EventTarget,
  all,
  allSettled,
  race,
  hash,
  hashSettled,
  rethrow,
  defer,
  denodeify,
  configure,
  on,
  off,
  resolve,
  reject,
  map,
  ['async']: async, // babel seems to error if async isn't a computed prop here...
  filter
};

export {
  asap,
  cast,
  Promise,
  EventTarget,
  all,
  allSettled,
  race,
  hash,
  hashSettled,
  rethrow,
  defer,
  denodeify,
  configure,
  on,
  off,
  resolve,
  reject,
  map,
  async,
  filter
};
